cmd/kubeadm/app/master/selfhosted.go

```diff
@@ -69,9 +87,271 @@ func CreateSelfHostedControlPlane(cfg *kubeadmapi.MasterConfiguration, client *c
 		return err
 	}
 
+	if err := launchSelfHostedProxy(cfg, client); err != nil {
+		return err
+	}
+
+	if err := launchEtcdOperator(cfg, client); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getEtcdTPRClient() (*rest.RESTClient, error) {
+	kubeConfigPath := path.Join(kubeadmapi.GlobalEnvParams.KubernetesDir, kubeadmconstants.AdminKubeConfigFileName)
+	config, err := clientcmd.BuildConfigFromFlags("", kubeConfigPath)
+	if err != nil {
+		return nil, err
+	}
+
+	config.GroupVersion = &schema.GroupVersion{
+		Group:   spec.TPRGroup,
+		Version: spec.TPRVersion,
+	}
+	config.APIPath = "/apis"
+	config.ContentType = runtime.ContentTypeJSON
+	config.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: api.Codecs}
+
+	restcli, err := rest.RESTClientFor(config)
+	if err != nil {
+		return nil, err
+	}
+	return restcli, nil
+}
+
+func getBootEtcdPodIP(kubecli *clientset.Clientset) (string, error) {
+	var ip string
+	err := wait.Poll(5*time.Second, 60*time.Second, func() (bool, error) {
+		podList, err := kubecli.CoreV1().Pods(api.NamespaceSystem).List(metav1.ListOptions{
+			LabelSelector: "component=boot-etcd",
+		})
+		if err != nil {
+			fmt.Printf("failed to list 'boot-etcd' pod: %v\n", err)
+			return false, err
+		}
+		if len(podList.Items) < 1 {
+			fmt.Printf("no 'boot-etcd' pod found, retrying after 5s...\n")
+			return false, nil
+		}
+
+		pod := podList.Items[0]
+		ip = pod.Status.PodIP
+		if len(ip) == 0 {
+			return false, nil
+		}
+		return true, nil
+	})
+	return ip, err
+}
+
+func CreateEtcdCluster(cfg *kubeadmapi.MasterConfiguration, client *clientset.Clientset) error {
+	start := time.Now()
+
+	// setup TPR client
+	restClient, err := getEtcdTPRClient()
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Wait for TPR to exist")
+	k8sutil.WaitEtcdTPRReady(restClient, time.Second*5, time.Minute*1, "kube-system")
+
+	seedPodIP, err := getBootEtcdPodIP(client)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("Boot IP is %s\n", seedPodIP)
+
+	// create TPR data
+	cluster := &spec.Cluster{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "etcd.coreos.com/v1beta1",
+			Kind:       "Cluster",
+		},
+		Metadata: metav1.ObjectMeta{
+			Name:      etcdCluster,
+			Namespace: metav1.NamespaceSystem,
+		},
+		Spec: spec.ClusterSpec{
+			Size:    1,
+			Version: "v3.1.6",
+			SelfHosted: &spec.SelfHostedPolicy{
+				BootMemberClientEndpoint: fmt.Sprintf("http://%s:12379", seedPodIP),
+			},
+			Pod: &spec.PodPolicy{
+				NodeSelector: map[string]string{"node-role.kubernetes.io/master": ""},
+				Tolerations: []v1.Toleration{
+					v1.Toleration{
+						Key:      "node-role.kubernetes.io/master",
+						Operator: "Exists",
+						Effect:   v1.TaintEffectNoSchedule,
```

Comment:
You don't need the value if the operator is Exists: https://github.com/kubernetes/kubernetes/blob/3a571384a2f117096f5602984f482fd79bc7a942/pkg/api/types.go#L2012